File: platform/how-to-get-started/2.fetch_httpx_gemini_openai_protocols/1.fetch-providers-gemini-openai-structure.py
import os
import json
import httpx

GEMINI_API_KEY = os.environ["GEMINI_API_KEY"]

url = "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions"

schema = {
    "type": "object",
    "properties": {
        "title": {"type": "string"},
        "priority": {"type": "string", "enum": ["low", "medium", "high"]},
        "tags": {"type": "array", "items": {"type": "string"}},
    },
    "required": ["title", "priority", "tags"],
    "additionalProperties": False,
}

payload = {
    "model": "gemini-3-flash-preview",
    "messages": [
        {"role": "system", "content": "Return ONLY valid JSON that matches the schema."},
        {"role": "user", "content": "Create a task for 'pay rent' with priority high and tags finance, home."},
    ],
    # OpenAI-style structured output request
    "response_format": {
        "type": "json_schema",
        "json_schema": {
            "name": "Task",
            "schema": schema,
            "strict": True
        }
    },
}

headers = {
    "Authorization": f"Bearer {GEMINI_API_KEY}",
    "Content-Type": "application/json",
}

with httpx.Client(timeout=30.0) as client:
    r = client.post(url, headers=headers, json=payload)
    r.raise_for_status()
    data = r.json()

content = data["choices"][0]["message"]["content"]
task = json.loads(content)

print(task)

File: platform/how-to-get-started/2.fetch_httpx_gemini_openai_protocols/1.fetch-providers-gemini-openai-structure.mjs
import { request } from "undici";

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) throw new Error("Missing GEMINI_API_KEY");

const url = "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions";

const schema = {
  type: "object",
  properties: {
    city: { type: "string" },
    tempF: { type: "number" },
    summary: { type: "string" },
  },
  required: ["city", "tempF", "summary"],
  additionalProperties: false,
};

const payload = {
  model: "gemini-3-flash-preview",
  messages: [
    { role: "system", content: "Return ONLY valid JSON matching the schema." },
    { role: "user", content: "Generate a sample weather report for Boston." },
  ],
  response_format: {
    type: "json_schema",
    json_schema: {
      name: "WeatherReport",
      schema,
      strict: true,
    },
  },
};

const { statusCode, body } = await request(url, {
  method: "POST",
  headers: {
    Authorization: `Bearer ${GEMINI_API_KEY}`,
    "Content-Type": "application/json",
  },
  body: JSON.stringify(payload),
});

const text = await body.text();
if (statusCode >= 400) throw new Error(`${statusCode} ${text}`);

const data = JSON.parse(text);
const content = data.choices?.[0]?.message?.content;
const report = JSON.parse(content);

console.log(report);
